staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go

```diff
@@ -0,0 +1,479 @@
+	return &streamProtocolV4{
+		StreamOptions: options,
+		binary:        true,
+	}
+}
+
+func newBase64V4(options StreamOptions) streamProtocolHandler {
+	return &streamProtocolV4{
+		StreamOptions: options,
+		binary:        false,
+	}
+}
+
+func (p *streamProtocolV4) copyStdin() {
+	if p.Stdin != nil {
+		var once sync.Once
+
+		// copy from client's stdin to container's stdin
+		/*go func() {
+			defer runtime.HandleCrash()
+
+			// if p.stdin is noninteractive, p.g. `echo abc | kubectl exec -i <pod> -- cat`, make sure
+			// we close remoteStdin as soon as the copy from p.stdin to remoteStdin finishes. Otherwise
+			// the executed command will remain running.
+			defer once.Do(func() { p.remoteStdinIn.Close() })
+
+			if _, err := io.Copy(p.remoteStdinOut, readerWrapper{p.Stdin}); err != nil {
+				runtime.HandleError(err)
+			}
+		}()*/
+
+		// read from remoteStdin until the stream is closed. this is essential to
+		// be able to exit interactive sessions cleanly and not leak goroutines or
+		// hang the client's terminal.
+		//
+		// TODO we aren't using go-dockerclient any more; revisit this to determine if it's still
+		// required by engine-api.
+		//
+		// go-dockerclient's current hijack implementation
+		// (https://github.com/fsouza/go-dockerclient/blob/89f3d56d93788dfe85f864a44f85d9738fca0670/client.go#L564)
+		// waits for all three streams (stdin/stdout/stderr) to finish copying
+		// before returning. When hijack finishes copying stdout/stderr, it calls
+		// Close() on its side of remoteStdin, which allows this copy to complete.
+		// When that happens, we must Close() on our side of remoteStdin, to
+		// allow the copy in hijack to complete, and hijack to return.
+		go func() {
+			defer runtime.HandleCrash()
+			defer once.Do(func() { p.remoteStdinIn.Close() })
+
+			// this "copy" doesn't actually read anything - it's just here to wait for
+			// the server to close remoteStdin.
+			if _, err := io.Copy(ioutil.Discard, p.remoteStdinIn); err != nil {
+				p.closeStreams()
+				runtime.HandleError(err)
+			}
+		}()
+	}
+}
+
+func (p *streamProtocolV4) copyStdout(wg *sync.WaitGroup) {
+	if p.Stdout == nil {
+		return
+	}
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
+		defer wg.Done()
+		defer p.closeStreams()
+
+		if _, err := io.Copy(p.Stdout, p.remoteStdoutIn); err != nil {
+			if err != io.EOF {
+				runtime.HandleError(err)
+			}
+		}
+	}()
+}
+
+func (p *streamProtocolV4) copyStderr(wg *sync.WaitGroup) {
+	if p.Stderr == nil || p.Tty {
+		return
+	}
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
+		defer wg.Done()
+		defer p.closeStreams()
+
+		if _, err := io.Copy(p.Stderr, p.remoteStderrIn); err != nil {
+			runtime.HandleError(err)
+		}
+	}()
+}
+
+func (p *streamProtocolV4) stream(conn *websocket.Conn) error {
+
+	defer conn.Close()
+	doneChan := make(chan struct{}, 2)
+
+	// set up error stream
+	p.errorStreamIn, p.errorStreamOut = io.Pipe()
```

Comment:
Why do we need these pipes? Can we stitch together the websocket connection with the streamOption's stdin/stdout/stderr directly?